File: frontend/src/lib/plugins/tests/permissions.test.ts
/**
 * Permission System Tests (Vitest Migration)
 * 
 * Tests the core permission management logic including pattern matching,
 * temporal permissions, and risk categorization using Vitest framework.
 */

import { describe, it, expect, beforeEach, vi } from 'vitest'
import { 
  PermissionManager, 
  getPermissionCategory, 
  getPermissionDescription,
  createPermissionRequest,
  PermissionCategory
} from '../runtime/permissions.js'
import type { Permission } from '../api.js'

describe('PermissionManager', () => {
  let manager: PermissionManager

  describe('initialization', () => {
    it('initializes with declared permissions', () => {
      const permissions: Permission[] = ['readRepo', 'writeRepo', 'secrets:jira*']
      manager = new PermissionManager(permissions)
      
      const declared = manager.getDeclaredPermissions()
      expect(declared).toHaveLength(3)
      expect(declared).toContain('readRepo')
      expect(declared).toContain('writeRepo')
      expect(declared).toContain('secrets:jira*')
    })

    it('denies undeclared permissions by default', () => {
      manager = new PermissionManager(['readRepo'])
      
      expect(manager.hasPermission('writeRepo')).toBe(false)
      expect(manager.hasPermission('net')).toBe(false)
    })
  })

  describe('permission granting and revoking', () => {
    beforeEach(() => {
      manager = new PermissionManager(['readRepo', 'writeRepo'])
    })

    it('grants declared permissions', () => {
      // Initially denied
      expect(manager.hasPermission('readRepo')).toBe(false)
      
      // Grant permission
      manager.grantPermission('readRepo')
      expect(manager.hasPermission('readRepo')).toBe(true)
      
      // Other permission still denied
      expect(manager.hasPermission('writeRepo')).toBe(false)
    })

    it('revokes permissions', () => {
      manager.grantPermission('readRepo')
      expect(manager.hasPermission('readRepo')).toBe(true)
      
      manager.revokePermission('readRepo')
      expect(manager.hasPermission('readRepo')).toBe(false)
    })

    it('provides granted permission summary', () => {
      // Initially no grants
      let grants = manager.getGrantedPermissions()
      expect(grants).toHaveLength(0)
      
      // Grant some permissions
      manager.grantPermission('readRepo')
      manager.grantPermission('writeRepo', { temporary: true, duration: 3600000 })
      
      grants = manager.getGrantedPermissions()
      expect(grants).toHaveLength(2)
      
      const readRepoGrant = grants.find(g => g.permission === 'readRepo')
      expect(readRepoGrant?.granted).toBe(true)
      expect(readRepoGrant?.temporary).toBeFalsy()
      
      const writeRepoGrant = grants.find(g => g.permission === 'writeRepo')
      expect(writeRepoGrant?.granted).toBe(true)
      expect(writeRepoGrant?.temporary).toBe(true)
      expect(writeRepoGrant?.expiresAt).toBeInstanceOf(Date)
    })
  })

  describe('temporary permissions', () => {
    beforeEach(() => {
      manager = new PermissionManager(['readRepo'])
    })

    it('handles temporary permissions with expiry', async () => {
      // Grant temporary permission for 50ms
      manager.grantPermission('readRepo', { temporary: true, duration: 50 })
      expect(manager.hasPermission('readRepo')).toBe(true)
      
      // Advance time by 60ms
      vi.advanceTimersByTime(60)
      
      // Should be expired now
      expect(manager.hasPermission('readRepo')).toBe(false)
    })

    it('tracks expiry dates correctly', () => {
      const before = new Date()
      manager.grantPermission('readRepo', { temporary: true, duration: 3600000 })
      const after = new Date(Date.now() + 3600000 + 1000) // Add 1s tolerance
      
      const grants = manager.getGrantedPermissions()
      const grant = grants.find(g => g.permission === 'readRepo')
      
      expect(grant?.expiresAt).toBeInstanceOf(Date)
      expect(grant?.expiresAt!.getTime()).toBeGreaterThan(before.getTime())
      expect(grant?.expiresAt!.getTime()).toBeLessThan(after.getTime())
    })
  })

  describe('pattern matching', () => {
    beforeEach(() => {
      manager = new PermissionManager(['secrets:jira*', 'secrets:github.token'])
    })

    it('matches wildcard patterns for secrets', () => {
      // Grant wildcard permission
      manager.grantPermission('secrets:jira*')
      
      // Should match specific jira secrets
      expect(manager.hasPermission('secrets:jira.token' as Permission)).toBe(true)
      expect(manager.hasPermission('secrets:jira.oauth.refresh' as Permission)).toBe(true)
      
      // Should not match other services
      expect(manager.hasPermission('secrets:github.token' as Permission)).toBe(false)
    })

    it('handles exact matches', () => {
      manager.grantPermission('secrets:github.token')
      expect(manager.hasPermission('secrets:github.token')).toBe(true)
    })

    it('handles edge cases in pattern matching', () => {
      const wildcardManager = new PermissionManager(['secrets:*'])
      wildcardManager.grantPermission('secrets:*' as Permission)
      
      // Wildcard should match everything under secrets
      expect(wildcardManager.hasPermission('secrets:anything' as Permission)).toBe(true)
      expect(wildcardManager.hasPermission('secrets:very.long.nested.key' as Permission)).toBe(true)
      
      // But not non-secrets
      expect(wildcardManager.hasPermission('readRepo')).toBe(false)
    })

    it('handles empty prefix patterns', () => {
      const emptyManager = new PermissionManager(['secrets:' as Permission])
      emptyManager.grantPermission('secrets:' as Permission)
      
      // Should only match exact empty suffix
      expect(emptyManager.hasPermission('secrets:' as Permission)).toBe(true)
      expect(emptyManager.hasPermission('secrets:something' as Permission)).toBe(false)
    })
  })

  describe('permission requests', () => {
    let mockCallback: ReturnType<typeof vi.fn>

    beforeEach(() => {
      manager = new PermissionManager(['readRepo', 'writeRepo'])
      mockCallback = vi.fn()
      manager.onPermissionRequest(mockCallback)
    })

    it('requests permission from callbacks', async () => {
      mockCallback.mockResolvedValue(true)
      
      const result = await manager.requestPermission({
        permission: 'readRepo',
        reason: 'Test reason'
      })
      
      expect(result).toBe(true)
      expect(mockCallback).toHaveBeenCalledWith({
        permission: 'readRepo',
        reason: 'Test reason'
      })
      expect(manager.hasPermission('readRepo')).toBe(true)
    })

    it('rejects undeclared permissions', async () => {
      await expect(manager.requestPermission({
        permission: 'net' // Not in declared permissions
      })).rejects.toThrow('Permission net not declared')
    })

    it('returns true if already granted', async () => {
      manager.grantPermission('readRepo')
      
      const result = await manager.requestPermission({
        permission: 'readRepo'
      })
      
      expect(result).toBe(true)
      expect(mockCallback).not.toHaveBeenCalled()
    })

    it('handles callback rejection', async () => {
      mockCallback.mockResolvedValue(false)
      
      const result = await manager.requestPermission({
        permission: 'readRepo'
      })
      
      expect(result).toBe(false)
      expect(manager.hasPermission('readRepo')).toBe(false)
    })
  })
})

describe('Permission utility functions', () => {
  describe('getPermissionCategory', () => {
    it('categorizes low risk permissions', () => {
      expect(getPermissionCategory('readRepo')).toBe(PermissionCategory.LOW_RISK)
      expect(getPermissionCategory('ui')).toBe(PermissionCategory.LOW_RISK)
    })

    it('categorizes medium risk permissions', () => {
      expect(getPermissionCategory('attachments')).toBe(PermissionCategory.MEDIUM_RISK)
      expect(getPermissionCategory('indexWrite')).toBe(PermissionCategory.MEDIUM_RISK)
    })

    it('categorizes high risk permissions', () => {
      expect(getPermissionCategory('writeRepo')).toBe(PermissionCategory.HIGH_RISK)
      expect(getPermissionCategory('net')).toBe(PermissionCategory.HIGH_RISK)
      expect(getPermissionCategory('secrets:jira*' as Permission)).toBe(PermissionCategory.HIGH_RISK)
    })
  })

  describe('getPermissionDescription', () => {
    it('provides human-readable descriptions', () => {
      const readDesc = getPermissionDescription('readRepo')
      expect(readDesc).toContain('Read project data')
      
      const writeDesc = getPermissionDescription('writeRepo')
      expect(writeDesc).toContain('Create, modify, and delete')
      
      const secretDesc = getPermissionDescription('secrets:jira*' as Permission)
      expect(secretDesc).toContain('Access stored secrets for jira')
    })

    it('handles unknown permissions', () => {
      const unknownDesc = getPermissionDescription('unknown:permission' as Permission)
      expect(unknownDesc).toContain('Unknown permission')
    })
  })

  describe('createPermissionRequest', () => {
    it('creates valid requests with defaults', () => {
      const request = createPermissionRequest('readRepo')
      
      expect(request.permission).toBe('readRepo')
      expect(request.reason).toContain('read your project data')
      expect(request.temporary).toBeUndefined()
      expect(request.duration).toBeUndefined()
    })

    it('accepts overrides', () => {
      const request = createPermissionRequest('writeRepo', {
        reason: 'Custom reason',
        temporary: true,
        duration: 1800000
      })
      
      expect(request.permission).toBe('writeRepo')
      expect(request.reason).toBe('Custom reason')
      expect(request.temporary).toBe(true)
      expect(request.duration).toBe(1800000)
    })

    it('handles secret permissions with fallback reason', () => {
      const request = createPermissionRequest('secrets:unknown' as Permission)
      expect(request.reason).toContain('stored authentication tokens')
    })
  })
})

File: internal/plugins/permissions_test.go
package plugins

import (
	"testing"
	"time"
)

func TestPermissionManager_DeclareAndCheck(t *testing.T) {
	pm := NewPermissionManager()
	pluginID := "com.test.plugin"

	// Test declaring permissions
	permissions := []Permission{PermissionReadRepo, PermissionWriteRepo}
	pm.DeclarePermissions(pluginID, permissions)

	declared := pm.GetDeclaredPermissions(pluginID)
	if len(declared) != 2 {
		t.Errorf("Expected 2 declared permissions, got %d", len(declared))
	}

	// Initially no permissions should be granted
	if pm.HasPermission(pluginID, PermissionReadRepo) {
		t.Error("Permission should not be granted initially")
	}
}

func TestPermissionManager_GrantPermission(t *testing.T) {
	pm := NewPermissionManager()
	pluginID := "com.test.plugin"
	
	// Declare permissions
	permissions := []Permission{PermissionReadRepo, PermissionWriteRepo}
	pm.DeclarePermissions(pluginID, permissions)

	// Test granting a declared permission
	envelope := pm.GrantPermission(pluginID, PermissionReadRepo, false, 0)
	if envelope.Code != "" {
		t.Errorf("Failed to grant permission: %v", envelope.Message)
	}

	if !pm.HasPermission(pluginID, PermissionReadRepo) {
		t.Error("Permission should be granted")
	}

	if pm.HasPermission(pluginID, PermissionWriteRepo) {
		t.Error("Ungranted permission should not be available")
	}

	// Test granting undeclared permission
	envelope = pm.GrantPermission(pluginID, PermissionNet, false, 0)
	if envelope.Code == "" {
		t.Error("Should not be able to grant undeclared permission")
	}
}

func TestPermissionManager_TemporaryPermissions(t *testing.T) {
	pm := NewPermissionManager()
	pluginID := "com.test.plugin"
	
	// Declare permissions
	permissions := []Permission{PermissionReadRepo}
	pm.DeclarePermissions(pluginID, permissions)

	// Grant temporary permission
	duration := 100 * time.Millisecond
	envelope := pm.GrantPermission(pluginID, PermissionReadRepo, true, duration)
	if envelope.Code != "" {
		t.Errorf("Failed to grant temporary permission: %v", envelope.Message)
	}

	// Should initially have permission
	if !pm.HasPermission(pluginID, PermissionReadRepo) {
		t.Error("Temporary permission should be available immediately")
	}

	// Wait for expiry
	time.Sleep(150 * time.Millisecond)

	// Should no longer have permission
	if pm.HasPermission(pluginID, PermissionReadRepo) {
		t.Error("Temporary permission should have expired")
	}
}

func TestPermissionManager_PatternMatching(t *testing.T) {
	pm := NewPermissionManager()
	pluginID := "com.test.plugin"
	
	// Declare wildcard permission
	permissions := []Permission{"secrets:jira*"}
	pm.DeclarePermissions(pluginID, permissions)

	// Grant wildcard permission
	envelope := pm.GrantPermission(pluginID, "secrets:jira*", false, 0)
	if envelope.Code != "" {
		t.Errorf("Failed to grant wildcard permission: %v", envelope.Message)
	}

	// Should match specific secrets
	if !pm.HasPermission(pluginID, "secrets:jira.token") {
		t.Error("Wildcard permission should match specific pattern")
	}

	if !pm.HasPermission(pluginID, "secrets:jira.oauth.refresh") {
		t.Error("Wildcard permission should match nested pattern")
	}

	// Should not match other services
	if pm.HasPermission(pluginID, "secrets:github.token") {
		t.Error("Wildcard permission should not match different service")
	}
}

func TestPermissionManager_RevokePermission(t *testing.T) {
	pm := NewPermissionManager()
	pluginID := "com.test.plugin"
	
	// Declare and grant permission
	permissions := []Permission{PermissionReadRepo}
	pm.DeclarePermissions(pluginID, permissions)
	
	envelope := pm.GrantPermission(pluginID, PermissionReadRepo, false, 0)
	if envelope.Code != "" {
		t.Errorf("Failed to grant permission: %v", envelope.Message)
	}

	// Verify permission is granted
	if !pm.HasPermission(pluginID, PermissionReadRepo) {
		t.Error("Permission should be granted")
	}

	// Revoke permission
	pm.RevokePermission(pluginID, PermissionReadRepo)

	// Verify permission is revoked
	if pm.HasPermission(pluginID, PermissionReadRepo) {
		t.Error("Permission should be revoked")
	}
}

func TestPermissionManager_PermissionCategories(t *testing.T) {
	pm := NewPermissionManager()

	tests := []struct {
		permission Permission
		category   string
	}{
		{PermissionReadRepo, "LOW_RISK"},
		{PermissionUI, "LOW_RISK"},
		{PermissionAttachments, "MEDIUM_RISK"},
		{PermissionIndexWrite, "MEDIUM_RISK"},
		{PermissionWriteRepo, "HIGH_RISK"},
		{PermissionNet, "HIGH_RISK"},
		{"secrets:jira.token", "HIGH_RISK"},
		{"secrets:github*", "HIGH_RISK"},
	}

	for _, tt := range tests {
		t.Run(string(tt.permission), func(t *testing.T) {
			category := pm.GetPermissionCategory(tt.permission)
			if category != tt.category {
				t.Errorf("Expected category %s, got %s", tt.category, category)
			}
		})
	}
}

func TestPermissionManager_PermissionDescriptions(t *testing.T) {
	pm := NewPermissionManager()

	tests := []struct {
		permission  Permission
		shouldContain string
	}{
		{PermissionReadRepo, "Read project data"},
		{PermissionWriteRepo, "Modify project data"},
		{PermissionAttachments, "Access file attachments"},
		{PermissionNet, "Access network resources"},
		{PermissionIndexWrite, "Write to search index"},
		{PermissionUI, "Contribute to user interface"},
		{"secrets:jira.token", "Access secrets for jira.token"},
		{"secrets:github*", "Access secrets for githubany"},
		{"unknown:permission", "Unknown permission"},
	}

	for _, tt := range tests {
		t.Run(string(tt.permission), func(t *testing.T) {
			description := pm.GetPermissionDescription(tt.permission)
			if description == "" {
				t.Error("Description should not be empty")
			}
			if tt.shouldContain != "" && !contains(description, tt.shouldContain) {
				t.Errorf("Description '%s' should contain '%s'", description, tt.shouldContain)
			}
		})
	}
}

func TestPermissionManager_CleanupExpiredGrants(t *testing.T) {
	pm := NewPermissionManager()
	pluginID := "com.test.plugin"
	
	// Declare permissions
	permissions := []Permission{PermissionReadRepo, PermissionWriteRepo}
	pm.DeclarePermissions(pluginID, permissions)

	// Grant one permanent and one temporary permission
	pm.GrantPermission(pluginID, PermissionReadRepo, false, 0) // permanent
	pm.GrantPermission(pluginID, PermissionWriteRepo, true, 50*time.Millisecond) // temporary

	// Both should be available initially
	if !pm.HasPermission(pluginID, PermissionReadRepo) {
		t.Error("Permanent permission should be available")
	}
	if !pm.HasPermission(pluginID, PermissionWriteRepo) {
		t.Error("Temporary permission should be available initially")
	}

	// Wait for temporary permission to expire
	time.Sleep(100 * time.Millisecond)

	// Clean up expired grants
	pm.CleanupExpiredGrants()

	// Permanent should still be available
	if !pm.HasPermission(pluginID, PermissionReadRepo) {
		t.Error("Permanent permission should still be available")
	}

	// Temporary should be cleaned up
	if pm.HasPermission(pluginID, PermissionWriteRepo) {
		t.Error("Expired temporary permission should be cleaned up")
	}
}

// Helper function to check if string contains substring
func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(s) > len(substr) && 
		(s[:len(substr)] == substr || s[len(s)-len(substr):] == substr || 
		 containsHelper(s, substr)))
}

func containsHelper(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}